### PR TITLE
Start pgbouncer in a proper daemonized mode

### PR DESCRIPTION
**Pull request by kigster**
# Description

This PR introduces the file `.profile.d/pgbouncer.sh`, which contains shared functions used by subsequent scripts to start `pgbouncer` or to stop it while waiting for the database connections to go away first.

Once again, we left the following scripts almost intact as to maintain backwards compatibility with the "old way" of running pgBouncer.

 * `bin/start-pgbouncer` (not used by us, but can be to revert to original buildpack behavior
 * `bin/use-server-pgbouncer` (also not used by us)

# How to use changes in the PR:

Instead of deciding whether to export variables in the Datadog's `prerun.sh` script, we are now doing that in this PR:

```bash
# decide if we want to print the connections to STDOUT
export PGBOUNCER_OUTPUT_URLS=true|false

[[ -x bin/start-pgbouncer-as-service ]] && {
  # Start pgBouncer as a daemon and the script to watch it and terminate it
  # when the dyno is shutting down
  bin/start-pgbouncer-as-service &
  
  # if PGBOUNCER_VARS_DISABLED is not set to "true", exports *_PGBOUNCER urlss
  source bin/use-client-pgbouncer
} 
```

# Shutdown

 * `SIGTERM` is ignored by the patched version of pgBouncer in this buildpack.
 * The script that starts pgBouncer as a daemon (`bin/start-pgbouncer-as-service`), itself starts an infinite loop, so that it can trap SIGTERM signal and begin the shutdown process. 
 * The shutdown process sends SIGINT to pgbouncer after all of the backend connections have been disconnected.

### CODE REVIEW

File: .profile.d/pgbouncer.sh [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+# Scripts in .profile.d/ should only be written by buildpacks. If you need to
+# perform application-specific initialization tasks at the time a dyno boots,
+# you should use .profile scripts, which are guaranteed to run after the
+# scripts in .profile.d/.
+
+export PGBOUNCER_COLOR_DISABLED=${PGBOUNCER_COLOR_DISABLED:-''}
+export PGBOUNCER_BINARY="vendor/pgbouncer/bin/pgbouncer"
+export PGBOUNCER_PROCESS_REGEX="vendor/pgbouncer/bin/[p]gbouncer"
+export PGBOUNCER_CONFIG_FILE="vendor/pgbouncer/pgbouncer.ini"
+export PGBOUNCER_POSTGRESQL_PORT=${PGBOUNCER_POSTGRESQL_PORT:-5432}
```
Comment (carlosramireziii) [inline .profile.d/pgbouncer.sh:29]: The buildpack starts PgBouncer on port 6000

File: .profile.d/pgbouncer.sh [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+# @description
+#   Returns the number of connections that pgbouncer has established
+#   to PostgresSQL
+function pgbouncer-current-connections-count() {
+  local pid="${1:-$(pgbouncer-get-current-pid)}"
+
+  if command -v lsof >/dev/null; then
+    lsof -p ${pid} -a -P -itcp | grep -c ":${PGBOUNCER_POSTGRESQL_PORT} (ESTABLISHED)"
+  elif command -v netstat >/dev/null; then
+    netstat -an | grep ESTABLISHED | grep -c ":${PGBOUNCER_POSTGRESQL_PORT}"
```
Comment (carlosramireziii) [inline .profile.d/pgbouncer.sh:151]: NOTE: `PGBOUNCER_POSTGRESQL_PORT` is set to 5432 by default but the buildpack starts it on 6000

File: bin/start-pgbouncer [hunk 3]
```diff
@@ -68,24 +71,21 @@ function run-pgbouncer() {
...
-  done
+# @description
+#   This function determines if the pgbouncer process should keep running
+#   because other dependencies might be using it.
+keep-running() {
+  declare pattern="$1"
+  declare -a pids
+  pids=($(pgrep -f "$pattern"))
+  [[ ${#pids[@]} -gt 0 ]]
 }
```
Comment (carlosramireziii) [inline bin/start-pgbouncer:82]: Is this function used anywhere?

File: bin/start-pgbouncer-as-service [hunk 4]
```diff
@@ -3,88 +3,7 @@
...
-      err "pgbouncer-stderr:"
-      cat "${stderr}"
-      rm -f "${stderr}"
-    }
-    return 1
-  fi
-}
-
-[[ "$0" != "${BASH_SOURCE[0]}" ]] || main "$@"
+pgbouncer-start-as-a-service
```
Comment (carlosramireziii) [inline bin/start-pgbouncer-as-service:9]: [nit] Does it make sense to define the function in the shared lib when it's only going to be called here?

File: .profile.d/pgbouncer.sh (hunk 2 above)
Comment (kigster) [reply .profile.d/pgbouncer.sh:151]: This is not looking for 6000. It's looking for the remote connection port to the actual DB. Basically the DB pool.

File: bin/use-client-pgbouncer [hunk 5]
```diff
@@ -65,6 +46,11 @@ function dbname() {
 function setup-client-pgbouncer() {
   local postgres_url
 
+  if is-enabled "${PGBOUNCER_VARS_DISABLED}"; then
+    inf "pgBouncer <VAR>_PGBOUNCER exports are disabled, skipping..."
+    return 0
+  fi
```
Comment (carlosramireziii) [inline bin/use-client-pgbouncer:52]: Note to selves: If this functionality is baked into the buildpack then we no longer need it in the application code

File: bin/start-pgbouncer (hunk 3 above)
Comment (kigster) [reply bin/start-pgbouncer:82]: Ah good catch.

File: .profile.d/pgbouncer.sh (hunk 2 above)
Comment (carlosramireziii) [reply .profile.d/pgbouncer.sh:151]: Ah gotcha! In that case WDYT of updating the name to distinguish that it's looking for the `PGBOUNCER_PG_SERVER_PORT` or something like that.

**Pull request closed without merging**